### ISSUE TITLE
eps prediction based on noisy x and t is [[[[Nan, Nan.....]]]] for 3 channels images

### ISSUE THREAD

**Issue by ishuiyutian** [joseph-nagel/diffusion-demo#1]
Hi @joseph-nagel , 

I'm learning the denoising diffusion model from the demo you published, I can train the ddpm mnist and ddpm swissroll successful on my computer. 

After that, I used another dataset which is a list of 64x64 images to train the ddpm model similarly as ddpm mnist, at the beginning, I convert the images to 1 channel images, and I can train it and show the trainning result by  `x_gen = ddpm.generate(sample_shape=image_shape, num_samples=25).cpu()`. 

When I try to train the model by 3 channels images, it can train the model without error, but when I want to generate the image by `x_gen = ddpm.generate(sample_shape=image_shape, num_samples=25).cpu()`, I can see after maybe 20 steps the model will generate the eps as something like: [[[[Nan, Nan.....]]]]

I'm trying to find out the reason but no result so far. That why I sent the message here, maybe you know more context about this. So, please help me find out the issue if you know more context.

At the end, thanks for you publishing this demo, I learn a lot from it.

Thanks

**Comment by joseph-nagel** [comment]
Hi,

The issue you are describing is a bit tricky to diagnose remotely. I have seen a certain diverging behavior during the sampling process in some other unrelated experiments with diffusion models as well. After all, there is a recursion going on, and some "instabilities" might lead to numerical overflows/underflows I guess.

Maybe you could try to let the model converge further in the training and observe whether the NaNs are still occuring. While the original parameter values have been roughly tuned for the two examples provided in the repository, there is actually no reason that the same hyperparameter values will work well for other problems, too. You can for example adapt the complexity of the model architecture. I have also observed that the results depend strongly on the chosen noise schedule. Hence, it would be worth checking whether changing the scheduling parameters would be of any help.

Just to mention, depending on your problem, your data and the standardization/normalization etc., there might be issues with bounds as well. The default formulation of DDPMs and the implementation in the code do not automatically respect or enforce the natural bounds of the data. I don't think that this is the reason for the NaNs, but it's good to still keep that in mind.

Let me know when there are any updates ...

**Comment by ishuiyutian** [comment]
Hi @joseph-nagel ,

Thanks for your help, I changed the mid_channels to `[64, 128, 256, 512]` and there are some checkpoints with low val loss that can generate the 3-channel image. Lets say I generate 9 images in a batch, there are about 3 images can be generated without problem and the other 6 still generate the Nan, I will test with more parameter values.
